packages/hardhat-chai-matchers/src/reverted/revertedWithCustomError.ts

```diff
@@ -0,0 +1,230 @@
+import { AssertionError } from "chai";
+import { decodeReturnData, getReturnDataFromError } from "./utils";
+
+const CUSTOM_ERROR_ASSERTION_CALLED = "customErrorAssertionCalled";
+const CUSTOM_ERROR_ASSERTION_DATA = "customErrorAssertionData";
+
+interface CustomErrorAssertionData {
+  contractInterface: any;
+  returnData: string;
+  customError: CustomError;
+}
+
+export function supportRevertedWithCustomError(
+  Assertion: Chai.AssertionStatic,
+  utils: Chai.ChaiUtils
+) {
+  Assertion.addMethod(
+    "revertedWithCustomError",
+    function (this: any, contract: any, expectedCustomErrorName: string) {
+      // check the case where users forget to pass the contract as the first
+      // argument
+      if (typeof contract === "string" || contract?.interface === undefined) {
+        throw new TypeError(
+          "The first argument of .revertedWithCustomError has to be the contract that defines the custom error"
+        );
+      }
+
+      // validate custom error name
+      if (typeof expectedCustomErrorName !== "string") {
+        throw new TypeError(
+          "Expected a string as the expected custom error name"
+        );
+      }
+
+      const iface: any = contract.interface;
+
+      const expectedCustomError = findCustomErrorByName(
+        iface,
+        expectedCustomErrorName
+      );
+
+      // check that interface contains the given custom error
+      if (expectedCustomError === undefined) {
+        throw new AssertionError(
```

Comment:
I feel like `AssertionError` doesn't make sense here, since it's not a failed user assertion that causes it to be emitted.

Reply:
Agree, fixed in 36be938fbb21e1b17a541199f4605c097d2f8aca This also made me realize that we didn't have a test for this scenario.